lib/transaction.js

```diff
@@ -0,0 +1,181 @@
+var debug = require('debug')('loopback:connector:transaction');
+var uuid = require('node-uuid');
+var utils = require('./utils');
+var jutil = require('./jutil');
+var ObserverMixin = require('./observer');
+
+var Transaction = require('loopback-connector').Transaction;
+
+module.exports = TransactionMixin;
+
+/**
+ * TransactionMixin class.  Use to add transaction APIs to a model class.
+ *
+ * @class TransactionMixin
+ */
+function TransactionMixin() {
+}
+
+/**
+ * Begin a new transaction
+ * @param {Object|String} [options] Options can be one of the forms:
+ * - Object: {isolationLevel: '...', timeout: 1000}
+ * - String: isolationLevel
+ *
+ * Valid values of `isolationLevel` are:
+ *
+ * - Transaction.READ_COMMITTED = 'READ COMMITTED'; // default
+ * - Transaction.READ_UNCOMMITTED = 'READ UNCOMMITTED';
+ * - Transaction.SERIALIZABLE = 'SERIALIZABLE';
+ * - Transaction.REPEATABLE_READ = 'REPEATABLE READ';
+ *
+ * @param {Function} cb Callback function. It calls back with (err, transaction).
+ * To pass the transaction context to one of the CRUD methods, use the `options`
+ * argument with `transaction` property, for example,
+ *
+ * ```js
+ *
+ * MyModel.beginTransaction('READ COMMITTED', function(err, tx) {
+ *   MyModel.create({x: 1, y: 'a'}, {transaction: tx}, function(err, inst) {
+ *     MyModel.find({x: 1}, {transaction: tx}, function(err, results) {
+ *       // ...
+ *       tx.commit(function(err) {...});
+ *     });
+ *   });
+ * });
+ * ```
+ *
+ * The transaction can be committed or rolled back. If timeout happens, the
+ * transaction will be rolled back. Please note a transaction is typically
+ * associated with a pooled connection. Committing or rolling back a transaction
+ * will release the connection back to the pool.
+ *
+ * Once the transaction is committed or rolled back, the connection property
+ * will be set to null to mark the transaction to be inactive. Trying to commit
+ * or rollback an inactive transaction will receive an error from the callback.
+ *
+ * Please also note that the transaction is only honored with the same data
+ * source/connector instance. CRUD methods will not join the current transaction
+ * if its model is not attached the same data source.
+ *
+ */
+TransactionMixin.beginTransaction = function(options, cb) {
+  cb = cb || utils.createPromiseCallback();
+  if (Transaction) {
```

Comment:
FWIW, if I am reading this code correctly, then `Transaction` is always set, because it is loaded from `loopback-connector` which is shipped as juggler's dependency. Should we check whether the backing connector (something like `Model.dataSource.connector`) supports transactions instead?